raiden/tasks.py

```diff
@@ -281,442 +439,793 @@ def _run(self):  # noqa pylint: disable=method-hidden,too-many-locals
             raiden.sign(mediated_transfer)
             forward_channel.register_transfer(mediated_transfer)
 
-            response = self.send_and_wait_valid(raiden, path, mediated_transfer)
-
-            # `next_hop` timedout
-            if response is None:
-                self.transfermanager.on_hashlock_result(hashlock, False)
+            for response in self.send_and_iter_valid(raiden, path, mediated_transfer):
+                valid_secretrequest = (
+                    isinstance(response, SecretRequest) and
+                    response.amount == amount and
+                    response.hashlock == hashlock and
+                    response.identifier == identifier
+                )
 
-            # someone down the line timedout / couldn't proceed
-            elif isinstance(response, (RefundTransfer, TransferTimeout)):
-                self.transfermanager.on_hashlock_result(hashlock, False)
+                if valid_secretrequest:
+                    # This node must reveal the Secret starting with the
+                    # end-of-chain, the `next_hop` can not be trusted to reveal the
+                    # secret to the other nodes.
+                    revealsecret_message = RevealSecret(secret)
+                    raiden.sign(revealsecret_message)
+
+                    # we cannot wait for ever since the `target` might
+                    # intentionally _not_ send the Ack, blocking us from
+                    # unlocking the asset.
+                    wait = (
+                        ESTIMATED_BLOCK_TIME * lock_timeout / .6
```

Comment:
> Hard coded constants are evil.
Agreed, I only needed a quick fix for not waiting until the lock timedout, I'm not sure if this is the best solution so I didn't bother with the `0.6` value.